keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1179 @@
+| --------------- | -------------------------------- | ------------------------------------------------------------------------------ |
+| Lock Type       | Lease                            | Lease                                                                          |
+| Claimed by      | Component instance               | Election Coordinator. (Lease is claimed for to the elected component instance) |
+| Renewed by      | Component instance               | Component instance                                                             |
+| Leader Criteria | First component to claim lease   | Best leader from available candidates at time of election                      |
+| Preemptable     | No                               | Yes, Collaboratively. (Coordinator marks lease as "end of term". Component instance voluntarily stops renewing) |
+
+### User Stories (Optional)
+
+#### Story 1
+
+A cluster administrator upgrades a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first and second nodes are upgraded, any components that were leaders
+  will typically lose the lease during the node downtime
+  - If one happens to retain its lease, it will be preempted by the coordinated
+    election controller after it updates its identity lease with new version
+    information
+- When the third node is upgraded, all components will be at the new version and
+  one will be elected
+
+#### Story 2
+
+A cluster administrator rolls back a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first node is rolled back, any components that were leaders will
+  typically loose the lease during the node downtime
+- Once one of the components updates its identity lease with new version
+  information, the coordinated election controller will preempt the current
+  leader so that this lower version component becomes leader.
+- When the remaining two nodes can rollback, the first node will typically
+  remain leader, but if a new election occurs, the available older version
+  components will be elected.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk: Amount of writes performed by leader election increases substantially
+
+This enhancement introduces an identity lease for each instance of each
+component.
+
+Example:
+
+- HA cluster with 3 control plane nodes
+- 3 elected components (kube-controller-manager, schedule,
+  cloud-controller-manager) per control plane node
+- 9 identity leases are created and renewed by the components
+
+Introducing this feature is roughtly equivalent to adding the same lease load as
+adding 9 nodes to a kubernetes cluster.
```

Comment:
This statement assumes that the "update period" is equivalent for node lease and for leader election. This is generally not true:
(a) we're using 10s for NodeLease
(b) we're using often 1s for leader-election

So it's more like adding 90 nodes - that's a huge difference.

I will just add one more data-point. If we assume few more components (most of providers actually have at least few other leader-elected components), then in relatively idle small cluster, even currently reducing leader-election renew interval from 1s to 10s is reducing the overall cpu load by 5%+. That's huge.
So I guess that introducing this mechanism would result in making this 15%+...

As long as I like this feature, I'm slightly worried that running even a small idle cluster is becoming more and more expensive and we're slowly moving towards a direction that even running a small idle cluster will require a beefy machine for the control-plane.
I would like us to explore a bit what we can do to optimize it. I'm not saying that it's the right model, but maybe a more interactive model like that could work:
- an instance creates its identity lease on start and refreshes it very infrequently (like once 1h or sh)
- when the controller is supposed to elect the leader it updates the AreYouStillAround bit on all identity leases
- in reaction to that instances update those identitiy leases
- elector wait up to say 5s for these to be updated (or if all are updated) and then chooses the new leader

That hugely reduces the load on these and rely more on watches.

Thoughts?

Reply:
That's a good point. Or even something along the lines of identity leases only updating themselves after seeing an end-of-term marker on the current lease.

I do see some potential caveats with the interactive model, though it's more of a tradeoff that we need to evaluate.
- we would have a slower transition time when a leader is lost (1s vs 5s or whatever we set the interval to).
- It's possible to be stuck in an infinite preemption loop if the best candidate for the lease is a slow writer (obviously an extreme edge case, but worth pointing out). The leader election controller will see that the best candidate did not update their identity lease in time for the new leader election and will elect one of the other candidates. After the candidate is selected and started, the best candidate updates their identity lease. The leader election controller will see that a better candidate is available and update the component lease with an `end-of-term` marker. After the lease expires the process is repeated. I think we would need to be careful in how we select how long to wait for the "acks" from the identity leases, but otherwise this seems like a sound approach

Alternatively, what's stopping us from using the same 10s interval for leader-election or an interval 1s < x < 10s?

Reply:
I fully agree with the interactive approach having drawbacks, as you pointed out. And I'm not attached to it - I just want to explore options to find something that is "cheaper".

Reducing leader-election interval means that in case of failure/crash/whatever, we stuck without the leader for a longer time. Also - operators do choose those interval themselves, so while we control defaults, we don't control what exactly will be used. So I don't think it's a real solution here.

Going back to the concerns of slow-writer - I think it will be problematic in general for leader-election. So I'm not entirely sure if the consequences of that mode aren't even more serious than you wrote, even for the default lgorithm.
But let's explore a slightly different one:
- the identity lease is updated very infrequently by the instance and has a long expiration (say 1m or sth like that)
- if a controller needs to elect new leader, it looks at all "healthy" ones [some of these may no longer be though]
- it chooses one, but then it's supposed to confirm it's healthiness quickly (in the next 5s or sth)
- if it doesn't confirm, new leader is elected soon after
In this case, the healthy case remains as it was before, in the unhelathy case (the newly elected leader in fact wasn't healthy), we have an additional turnaround, but that should be fairly rare situation...

Reply:
[Again, I'm not saying this one is the right model either - all I'm trying to do is to brainstorm ideas and try to figure out something that will be "cheaper"]

Reply:
> Reducing leader-election interval means that in case of failure/crash/whatever, we stuck without the leader for a longer time. Also - operators do choose those interval themselves, so while we control defaults, we don't control what exactly will be used. 

Introducing coordinated leader election creates two configurable intervals:
- Leader election renewal interval (For a candidate that has already been selected as leader to continue leading)
- Candidate lease renewal interval (For candidates to advertise they're available)

The concern with the high amount of writes is attached to the candidate lease renewal interval. One proposal is to increase candidate lease renewal interval while maintaining a low leader election renewal interval.

I agree that slow-writer is a problem in general. The interactive approach is quite promising, I wanted to point out the drawbacks so we have an awareness of them.

>the identity lease is updated very infrequently by the instance and has a long expiration (say 1m or sth like that)
 +1, note that even without the interactive approach, maintaining the leader lease to have a lower refresh time allows us to transition quick fast. 

I talked to Joe about this and the interactive approach is a good way to reduce write overhead. Will update the KEP to reflect this new approach.